rfcs/2020-11-01-remap-metrics-support.md

```diff
@@ -0,0 +1,191 @@
+simplify their configurations if they already use the `*_tag` transforms
+provided. It will also allow them remove the use of the `Lua` transform if they
+have been falling back to that for anything but the most complex manipulation.
+
+  
+## Prior Art
+
+
+## Drawbacks
+
+In order to maintain these functions there will be a slight maintenance burden.
+
+By exposing the metric data to Remap, it may place some restrictions should we 
+want to change the internal model in the future. The more we decide to expose to
+the language, the more it may restrict us going forwards if we need to ensure
+backwards compatibility for users configurations.
+
+## Alternatives
+
+### Use paths
+
+We could allow the metric data to be manipulated through paths in a similar way
+to how the language works with log events.
+
+The following fields would be available:
+
+- `.name` - The name of the metric.
+- `.namespace` - The namespace for the metric.
+- `.timestamp` - Timestamp of the metric.
+- `.kind` - The kind of the metric - *Incremental* or *Absolute*
+- `.tags` - A map containing the tags set for the metric.
+ 
+However, there are strict limitations to how this data can be represented. 
+Setting `name` or `namespace` to anything other than a string should be an 
+error. `tags` has to be a map of string keys to string values. No nesting should
+be allowed. The inflexibility around this data means that the attempt to make
+it "feel" the same as log events would not be effective.
+
+### Manipulate the metric data
+
+We could provide functions to allow manipulating the metric data itself.
+
+This is complicated because there are a number of different metric types,
+so each provided function would only work if the metric was the correct type.
+For example, functions `get_gauge_value` and `set_gauge_value` would error
+if the metric was an `AggregatedHistogram`. Similar problems would occur
+if the data was represented through paths rather than functions.
+
+This functionality is provided by the `Lua` transform, so there is still a way
+forward for the user if they need this functionality.
+
+Note that going with the original proposal does not preclude this, so this could
+still be implemented later as a next stage.
+
+### Allow conversion between metric and log events
+
+Currently the RFC just lays out methods to manipulate either log or metric 
+events. It could also be possible to add functions to convert a log event into
+a metric event and viceversa.
+
+For example:
+
+`. = to_log()`
+
+could take the current metric event and write the fields into the root path to 
+create a log event out of the metric.
+
+To create metric events one function per metric type could be provided such as:
+`make_counter(value)`, `make_gauge(value)`, `make_set([value])`
+
+The other functions provided in this [proposal](#doc-level-Proposal) can then be
+used to fill out the rest of the event.
+
+Note this could still be implemented at a later stage.
```

Comment:
This would be nice, and I agree it would be best to tackle this once we have more experience with metrics in Remap.